k8s-operator/apis/v1alpha1/types_proxyclass.go

```diff
@@ -188,8 +188,34 @@ type Metrics struct {
 type ServiceMonitor struct {
 	// If Enable is set to true, a Prometheus ServiceMonitor will be created. Enable can only be set to true if metrics are enabled.
 	Enable bool `json:"enable"`
+	// Labels to add to the ServiceMonitor.
+	// Labels must be valid Kubernetes labels.
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
+	// +optional
+	Labels Labels `json:"labels"`
+}
+
+type Labels map[string]LabelValue
+
+func (l Labels) Parse() map[string]string {
+	if l == nil {
+		return nil
+	}
+	m := make(map[string]string, len(l))
+	for k, v := range l {
+		m[k] = string(v)
+	}
+	return m
 }
 
+// We do not validate the values of the label keys here - it is done by the ProxyClass
+// reconciler because the validation rules are too complex for a CRD validation markers regex.
+
+// +kubebuilder:validation:Type=string
+// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9][-._a-zA-Z0-9]*[a-zA-Z0-9]){0,1}$`
```

Comment:
uber nit: I haven't tested, but I think this edit would allow a single-character value that wouldn't otherwise be allowed

```suggestion
// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]([-._a-zA-Z0-9]*[a-zA-Z0-9])?){0,1}$`
```

Reply:
Thanks for catching the single character issue
That edit would not allow an empty value which is also legit

I've changed this to the exact regex that kube uses
`^(([a-zA-Z0-9][-._a-zA-Z0-9]*)?[a-zA-Z0-9])?$` which allows both empty and single char values
https://github.com/kubernetes/apimachinery/blob/3e8e52d6a1259ada73f63c1c7d1fad39d4ba9fb4/pkg/util/validation/validation.go#L155